File: examples2d/one_way_platforms2.rs
use rapier_testbed2d::Testbed;
use rapier2d::prelude::*;

struct OneWayPlatformHook {
    platform1: ColliderHandle,
    platform2: ColliderHandle,
}

impl PhysicsHooks for OneWayPlatformHook {
    fn modify_solver_contacts(&self, context: &mut ContactModificationContext) {
        // The allowed normal for the first platform is its local +y axis, and the
        // allowed normal for the second platform is its local -y axis.
        //
        // Now we have to be careful because the `manifold.local_n1` normal points
        // toward the outside of the shape of `context.co1`. So we need to flip the
        // allowed normal direction if the platform is in `context.collider_handle2`.
        //
        // Therefore:
        // - If context.collider_handle1 == self.platform1 then the allowed normal is +y.
        // - If context.collider_handle2 == self.platform1 then the allowed normal is -y.
        // - If context.collider_handle1 == self.platform2 then its allowed normal +y needs to be flipped to -y.
        // - If context.collider_handle2 == self.platform2 then the allowed normal -y needs to be flipped to +y.
        let mut allowed_local_n1 = Vector::zeros();

        if context.collider1 == self.platform1 {
            allowed_local_n1 = Vector::y();
        } else if context.collider2 == self.platform1 {
            // Flip the allowed direction.
            allowed_local_n1 = -Vector::y();
        }

        if context.collider1 == self.platform2 {
            allowed_local_n1 = -Vector::y();
        } else if context.collider2 == self.platform2 {
            // Flip the allowed direction.
            allowed_local_n1 = Vector::y();
        }

        // Call the helper function that simulates one-way platforms.
        context.update_as_oneway_platform(&allowed_local_n1, 0.1);

        // Set the surface velocity of the accepted contacts.
        let tangent_velocity =
            if context.collider1 == self.platform1 || context.collider2 == self.platform2 {
                -12.0
            } else {
                12.0
            };

        for contact in context.solver_contacts.iter_mut() {
            contact.tangent_velocity.x = tangent_velocity;
        }
    }
}

pub fn init_world(testbed: &mut Testbed) {
    /*
     * World
     */
    let mut bodies = RigidBodySet::new();
    let mut colliders = ColliderSet::new();
    let impulse_joints = ImpulseJointSet::new();
    let multibody_joints = MultibodyJointSet::new();

    /*
     * Ground
     */
    let rigid_body = RigidBodyBuilder::fixed();
    let handle = bodies.insert(rigid_body);

    let collider = ColliderBuilder::cuboid(25.0, 0.5)
        .translation(vector![30.0, 2.0])
        .active_hooks(ActiveHooks::MODIFY_SOLVER_CONTACTS);
    let platform1 = colliders.insert_with_parent(collider, handle, &mut bodies);
    let collider = ColliderBuilder::cuboid(25.0, 0.5)
        .translation(vector![-30.0, -2.0])
        .active_hooks(ActiveHooks::MODIFY_SOLVER_CONTACTS);
    let platform2 = colliders.insert_with_parent(collider, handle, &mut bodies);

    /*
     * Setup the one-way platform hook.
     */
    let physics_hooks = OneWayPlatformHook {
        platform1,
        platform2,
    };

    /*
     * Spawn cubes at regular intervals and apply a custom gravity
     * depending on their position.
     */
    testbed.add_callback(move |graphics, physics, _, run_state| {
        if run_state.timestep_id % 200 == 0 && physics.bodies.len() <= 7 {
            // Spawn a new cube.
            let collider = ColliderBuilder::cuboid(1.5, 2.0);
            let body = RigidBodyBuilder::dynamic().translation(vector![20.0, 10.0]);
            let handle = physics.bodies.insert(body);
            physics
                .colliders
                .insert_with_parent(collider, handle, &mut physics.bodies);

            if let Some(graphics) = graphics {
                graphics.add_body(handle, &physics.bodies, &physics.colliders);
            }
        }

        for handle in physics.islands.active_dynamic_bodies() {
            let body = &mut physics.bodies[*handle];
            if body.position().translation.y > 1.0 {
                body.set_gravity_scale(1.0, false);
            } else if body.position().translation.y < -1.0 {
                body.set_gravity_scale(-1.0, false);
            }
        }
    });

    /*
     * Set up the testbed.
     */
    testbed.set_world_with_params(
        bodies,
        colliders,
        impulse_joints,
        multibody_joints,
        vector![0.0, -9.81],
        physics_hooks,
    );
    testbed.look_at(point![0.0, 0.0], 20.0);
}

File: python/rapier-testbed/rapier_testbed/examples2/one_way_platforms2.py
"""Port of examples2d/one_way_platforms2.rs."""
from __future__ import annotations

import rapier2d as rp
from .._registry import register

CATEGORY = "Collisions"
NAME = "One-way platforms"


class OneWayPlatformHook:
    """PhysicsHook that makes two platforms only allow contacts from one side."""

    def __init__(self, platform1, platform2) -> None:
        self.platform1 = platform1
        self.platform2 = platform2

    def modify_solver_contacts(self, ctx) -> None:
        allowed_local_n1 = (0.0, 0.0)

        if ctx.collider1 == self.platform1:
            allowed_local_n1 = (0.0, 1.0)
        elif ctx.collider2 == self.platform1:
            allowed_local_n1 = (0.0, -1.0)

        if ctx.collider1 == self.platform2:
            allowed_local_n1 = (0.0, -1.0)
        elif ctx.collider2 == self.platform2:
            allowed_local_n1 = (0.0, 1.0)

        ctx.update_as_oneway_platform(allowed_local_n1, 0.1)

        tangent_velocity = -12.0 if (
            ctx.collider1 == self.platform1 or ctx.collider2 == self.platform2
        ) else 12.0
        for i in range(ctx.num_solver_contacts()):
            sc = ctx.solver_contacts[i]
            try:
                sc.tangent_velocity = (tangent_velocity, sc.tangent_velocity.y)
            except Exception:
                # Some bindings expose tangent_velocity as a scalar; skip if
                # the assignment shape doesn't match.
                pass


def init_world(testbed) -> None:
    bodies = rp.RigidBodySet()
    colliders = rp.ColliderSet()
    impulse_joints = rp.ImpulseJointSet()
    multibody_joints = rp.MultibodyJointSet()

    handle = bodies.insert(rp.RigidBody.fixed())
    platform1 = colliders.insert_with_parent(
        rp.Collider.cuboid(25.0, 0.5)
        .translation((30.0, 2.0))
        .active_hooks(rp.ActiveHooks.MODIFY_SOLVER_CONTACTS),
        handle,
        bodies,
    )
    platform2 = colliders.insert_with_parent(
        rp.Collider.cuboid(25.0, 0.5)
        .translation((-30.0, -2.0))
        .active_hooks(rp.ActiveHooks.MODIFY_SOLVER_CONTACTS),
        handle,
        bodies,
    )

    hooks = OneWayPlatformHook(platform1, platform2)

    state = {"step": 0}

    def spawn_cube_callback(tb) -> None:
        state["step"] += 1
        if state["step"] % 200 == 0 and len(tb.bodies) <= 7:
            h = tb.bodies.insert(
                rp.RigidBody.dynamic().translation((20.0, 10.0))
            )
            tb.colliders.insert_with_parent(
                rp.Collider.cuboid(1.5, 2.0), h, tb.bodies
            )

        for h in tb._islands.active_dynamic_set():
            rb = tb.bodies.get(h)
            if rb is None:
                continue
            if rb.translation.y > 1.0:
                rb.gravity_scale = 1.0
                tb.bodies.replace(h, rb)
            elif rb.translation.y < -1.0:
                rb.gravity_scale = -1.0
                tb.bodies.replace(h, rb)

    testbed.add_callback(spawn_cube_callback)
    testbed.set_world_with_params(
        bodies, colliders, impulse_joints, multibody_joints, (0.0, -9.81), hooks
    )
    testbed.set_camera_2d(center=(0.0, 0.0), zoom=20.0)


register(CATEGORY, NAME, init_world, dim=2)

if __name__ == "__main__":
    from .._picker import run

    run(initial=f"{CATEGORY} / {NAME}")
